libraries/AP_Compass/AP_Compass_AK8963.cpp

```diff
@@ -237,30 +237,33 @@ void AP_Compass_AK8963::_make_factory_sensitivity_adjustment(Vector3f& field) co
 
 void AP_Compass_AK8963::_update()
 {
+    struct AP_AK8963_SerialBus::raw_value rv;
+    float mag_x, mag_y, mag_z;
+
     if (hal.scheduler->micros() - _last_update_timestamp < 10000) {
-        return;
+        goto end;
     }
 
     if (!_sem_take_nonblocking()) {
-        return;
+        goto end;
     }
 
-    struct AP_AK8963_SerialBus::raw_value rv;
-
     _bus->read_raw(&rv);
 
     /* Check for overflow. See AK8963's datasheet, section
      * 6.4.3.6 - Magnetic Sensor Overflow. */
     if ((rv.st2 & 0x08)) {
-        return;
+        _sem_give();
+        goto fail;
     }
 
-    float mag_x = (float) rv.val[0];
-    float mag_y = (float) rv.val[1];
-    float mag_z = (float) rv.val[2];
+    mag_x = (float) rv.val[0];
+    mag_y = (float) rv.val[1];
+    mag_z = (float) rv.val[2];
 
     if (is_zero(mag_x) && is_zero(mag_y) && is_zero(mag_z)) {
-        return;
+        _sem_give();
```

Comment:
Mistake,

Thank you,

Julien

2015-08-03 10:43 GMT+02:00 Tom Pittenger <notifications@github.com>:

> In libraries/AP_Compass/AP_Compass_AK8963.cpp
> <https://github.com/diydrones/ardupilot/pull/2646#discussion_r36065716>:
>
> >
> >      if (is_zero(mag_x) && is_zero(mag_y) && is_zero(mag_z)) {
> > -        return;
> > +        _sem_give();
>
> Why give twice?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2646/files#r36065716>.
>